feat(paths): create class to combine all qTox managed paths

sudden6 opened the pull request:
This will be the central location for all of qTox managed directories. It implements the directory structure as discussed by https://github.com/qTox/qTox/issues/5410

- [ ] Commits follow our [git commit guidelines](https://github.com/qTox/qTox/blob/master/CONTRIBUTING.md#git-commit-guidelines)

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/qtox/qtox/5427)
<!-- Reviewable:end -->

sudden6 committed:
feat(paths): create class to combine all qTox managed paths

This will be the central location for all of qTox managed directories.

sudden6 commented:
@anthonybilinski or @sphaerophoria please review.

Note that the upgrade logic as well as the switch to these paths will follow in another PR.

Diadlo commented on src/persistence/paths.cpp:
```diff
@@ -0,0 +1,243 @@
+ * @note forcePortable and forceNonPortable can't be true at the same time
+ *       if both of them are false, we autodetect if portable mode should be used
+ */
+Paths* Paths::makePaths(bool forcePortable, bool forceNonPortable)
```
Two conflicting bools in parameters looks strange. What about enum?
`enum Path::Portable { Auto, Portable, NonPortable };`

sudden6 replied:
I thought about it, but decided for standard types, will change.

Diadlo commented on src/persistence/paths.cpp:
```diff
@@ -0,0 +1,243 @@
+        return nullptr;
+    }
+
+    result->portable = portable;
```
Could we pass both in constructor?

sudden6 replied:
Done

Diadlo reviewed (commented):
What about some tests?

sudden6 commented:
> What about some tests?

I have no idea how to test this reliably, any suggestions?

Diadlo commented:
@sudden6 Create `Paths` with different values of `force[Non]Portable` and check that result corresponds to correct. Also check portable autodetection

Diadlo commented on src/persistence/paths.cpp:
```diff
@@ -0,0 +1,234 @@
+    return new Paths(basePath, portable);
+}
+
+Paths::Paths(QString basePath, bool portable)
```
Const reference for `QString`. Otherwise LGTM

sudden6 replied:
Done

sudden6 commented:
> Create Paths with different values of force[Non]Portable and check that result corresponds to correct

The problem with that is, that the paths depend on the system, ie they could be different on Travis CI vs my computer. Or are you suggesting to test against QStandardPaths?

Diadlo commented:
You are right, I missed it. But yes, tests can use `QStandartPath` too

sudden6 committed:
test(paths): add tests for paths module

sudden6 committed:
fix(paths): fix bugs uncovered while developing test cases

There was one directory separator too much in the path. Also improved
the documentation.

sudden6 commented:
@Diadlo Did tests

Diadlo commented on test/persistence/paths_test.cpp:
```diff
@@ -0,0 +1,168 @@
+    // auto detection should succeed
+    QVERIFY(paths != nullptr);
+    // the test environment should not provide a `qtox.ini`
+    QVERIFY(paths->isPortable() == false);
```
Suggest to avoid `bool == false` and `== true`. Feel free to ignore

sudden6 replied:
Especially in tests I prefer to make this more explicit.

Diadlo commented on test/persistence/paths_test.cpp:
```diff
@@ -0,0 +1,168 @@
+    const QString curPath{qApp->applicationDirPath()};
+    QStringList themeFolders{curPath % sep % themeFolder % sep};
+
+    QVERIFY(paths->getThemeDirs() == themeFolders);
```
Why not `verifyqToxPath`?

sudden6 replied:
`getThemeDirs` returns a `QStringList` this doesn't fit `verifyqToxPath`

Diadlo commented:
@sudden6 Sorry, forgot to click "publish comment" :/

sudden6 commented:
@Diadlo all updated

btw, Tox some time?

Diadlo reviewed (approved):
<img class="emoji" title=":lgtm_strong:" alt=":lgtm_strong:" align="absmiddle" src="https://reviewable.io/lgtm_strong.png" height="20" width="61"/>

I will try to be online soon :)

> Reviewed 1 of 3 files at r1, 2 of 3 files at r3, 2 of 2 files at r4.
> *[Reviewable](https://reviewable.io/reviews/qtox/qtox/5427#-:-LRFZqLJDQnQdQ3rVgvQ:bgrikck)* status: :shipit: complete! 1 of 1 LGTMs obtained



<!-- Sent from Reviewable.io -->
